File: 0383-ransom-note/0383-ransom-note.cpp
class Solution {
public:
    bool canConstruct(std::string ransomNote, std::string magazine) {
        // Create a hashmap to store the frequencies of characters in the magazine
        std::unordered_map<char, int> magazine_map;

        // Populate the magazine_map with the character frequencies
        for (char c : magazine) {
            magazine_map[c]++;
        }

        // Check if all characters in ransomNote can be found in magazine_map with sufficient frequency
        for (char c : ransomNote) {
            // If the character is not found or its frequency is zero, return false
            if (magazine_map[c] == 0) {
                return false;
            }
            // Decrease the frequency of the character in magazine_map
            magazine_map[c]--;
        }

        // If all characters were found with sufficient frequency, return true
        return true;
    }
};

File: 0383-ransom-note/0383-ransom-note.py
class Solution:
    def canConstruct(self, ransomNote: str, magazine: str) -> bool:
        # Hashmaps to store character frequencies
        ransom_hashmap = {}
        magazine_hashmap = {}

        # Populate the ransom_hashmap with the frequencies of characters in ransomNote
        for char in ransomNote:
            if char in ransom_hashmap:
                ransom_hashmap[char] += 1
            else:
                ransom_hashmap[char] = 1

        # Populate the magazine_hashmap with the frequencies of characters in magazine
        for char in magazine:
            if char in magazine_hashmap:
                magazine_hashmap[char] += 1
            else:
                magazine_hashmap[char] = 1

        # Check if the magazine can provide the characters needed for the ransom note
        for key, value in ransom_hashmap.items():
            if magazine_hashmap.get(key, 0) < value:
                return False

        return True
